File: code/sequence.kt
// Sequence is lazy
val seq = sequenceOf(1, 2, 3, 4)
    .filter { println("Filter $it, "); it % 2 == 1 }
    .map { println("Map $it, "); it * 2 }

// Computations are evaluated during terminal operation
val items = seq.toList()

// Filter 1,
// Map 1,
// Filter 2,
// Filter 3,
// Map 3,
// Filter 4,

// List is not lazy, so functions are evaluated immediately
val items2 = listOf(1, 2, 3, 4)
    .filter { println("Filter $it, "); it % 2 == 1 }
    .map { println("Map $it, "); it * 2 }

// Filter 1,
// Filter 2,
// Filter 3,
// Filter 4,
// Map 1,
// Map 3,

File: code/sequence.cs
// LINQ is lazy, so no need to use other collection types
var query = new List&lt;int> { 1, 2, 3, 4 }
        .Where(i =>
        {
            Console.WriteLine($"Filter {i}, ");
            return i % 2 == 1;
        }).Select(i =>
        {
            Console.WriteLine($"Map {i}, ");
            return i * 2;
        });


var items = query.ToList();

//Filter 1,
//Map 1,
//Filter 2,
//Filter 3,
//Map 3,
//Filter 4,